Fixed #31811 -- Add optional timing outputs to the test runner

lothemar opened the pull request:
@carltongibson While this does the job, I'm not sure if this is efficient.

Any thoughts about the TimeKeeper class?

ticket-31811

carltongibson commented:
Oh, can I ask you to fix the isort error? (Otherwise I get a twitch...) Ta!

lothemar commented:
Fixed the isort error (sorry for the twitch)

orf commented on django/test/utils.py:
```diff
@@ -841,6 +843,40 @@ def disable(self):
+    def results(self):
+        results = ""
+        for record in self._record:
+            results += "%s\n" % record
```
`results = "\n".join(self._record)` I think?

orf commented on django/test/utils.py:
```diff
@@ -841,6 +843,40 @@ def disable(self):
+time_keeper = TimeKeeper()
+
+
+class time_it(TestContextDecorator):
```
I would personally structure this a bit differently by combining `time_keeper` and `time_it`:

```python
class TimeKeeper:
    def __init__(...):
        self.records = []

   @contextlib.contextmanager
   def timed(self, name):
      t1 = time.time()
      yield
      self.records.append(...)

   def __iter__(self):
      yield from records
```

Then you can do:

```python
keeper = TimeKeeper()

with keeper.time("test"):
    ...

for line in keeper:
    print(line)
```

or something akin to that, which feels more natural.

adamchainz replied:
Yes this would be a good idea

lothemar replied:
Yup, this is exactly what I felt I was missing when I made the initial commit.

adamchainz commented on django/test/utils.py:
```diff
@@ -174,12 +175,13 @@ def setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, paral
-                            verbosity=verbosity,
-                            keepdb=keepdb,
-                        )
+                    with time_it("Database cloning took", timing):
```
```suggestion
                    with time_it("Database cloning", timing):
```

carltongibson commented:
> Could maybe be done in a follow-up.

Can I vote for this? (I'd like to get this in so we can start using it for the parallelisation PRs) (Unless it's going to be quick, in which case ... 😀)

lothemar commented:
@carltongibson Let me know if there's anything I need to change. This should be good-to-go for merging.

@orf I combined your commit with time_keeper to retain a record of all tests run. Is there anything 'unnatural' about the code as of now?

I'm overriding a private method in unittest.TextTestRunner to inject time_keeper in. Is this risky in the long-term? If so, I'll try to rewrite it

orf commented:
No, this looks good to me @lothemar! Great work. 

I will just note that this won't work in Jenkins because it uses the [XMLRunner](https://github.com/xmlrunner/unittest-xml-reporting) module. Unittest is a bit of a mess and is not very composable.

Lets merge this 🥳

felixxm commented on django/test/utils.py:
```diff
@@ -841,6 +843,43 @@ def disable(self):
+        if record[:4] == 'test':
+            self.test_suite_records[record] = record_time
+        else:
+            new_record = "%s took %0.3fs" % (record, record_time)
```
```suggestion
            new_record = '%s took %.3fs' % (record, record_time)
```

felixxm commented on django/test/runner.py:
```diff
@@ -424,6 +426,51 @@ def __iter__(self):
+        total = time.perf_counter() - self.start_time
+        self.start_time = None
+        if self.showAll:
+            self.stream.write("{:.3f}s ".format(total))
```
```suggestion
            self.stream.write('%.3fs ' % total))
```
Do we need a trailing space?

lothemar replied:
No, we shouldn't need it. The stream gets flushed after every test

smithdc1 commented on django/test/runner.py:
```diff
@@ -437,7 +484,8 @@ def __init__(self, pattern=None, top_level=None, verbosity=1,
                  tags=None, exclude_tags=None, test_name_patterns=None,
-                 pdb=False, buffer=False, enable_faulthandler=True, **kwargs):
+                 pdb=False, buffer=False, enable_faulthandler=True,
+                 timing=-1, **kwargs):
```
Pytest uses `durations` rather than timing. Any benefit in aligning? 

https://docs.pytest.org/en/stable/usage.html#profiling-test-execution-duration

adamchainz replied:
I think there is, just so folks have a little less to learn between the two.

lothemar replied:
It would be nice to align it, considering we've aligned similar pytest features.

felixxm commented on django/test/utils.py:
```diff
@@ -841,6 +843,43 @@ def disable(self):
+            results.append("")
+            results.extend(self.records)
+            results = '\n'.join(results)
+            print(results)
```
I would use `sys.stderr.write(msg + os.linesep)` for each result instead of `print()` with joined results.

lothemar commented:
@felixxm Thank you for your comments! 

Sorry for the confusing wording of the docs. I'll add a patch for database teardown, I hadn't considered it in the timings.

Currently if you pass ```--durations=0```, it'll show you **only** the timing outputs such as database creation, database cloning, and total run time, so you can only see these if you don't want to know your N-slowest tests. Passing any value greater than 0 will show you the timing outputs **and** N-slowest tests. Similarly, neither the test times aren't collected nor the ```TextTest*``` subclasses are used unless ```durations > 0```.

lothemar commented on django/test/utils.py:
```diff
@@ -841,6 +843,43 @@ def disable(self):
+                    results.append("%s took %0.3fs" % (key, value))
+                    self.n_tests -= 1
+        if self.n_tests >= 0:
+            results.append("")
```
For formatting so that timing outputs such as database cloning and teardown are visually separated from the N-slowest tests

felixxm commented:
> Currently if you pass `--durations=0`, it'll show you **only** the timing outputs such as database creation, database cloning, and total run time, so you can only see these if you don't want to know your N-slowest tests. Passing any value greater than 0 will show you the timing outputs **and** N-slowest tests. Similarly, neither the test times aren't collected nor the `TextTest*` subclasses are used unless `durations > 0`.

Yes I know, IMO this is confusing, unexpected and out of scope of this ticket (grumbling ... :older_man:), maybe it's clear for others ... :shrug: , if I want to get timing outputs I have to declare the number of slowest tests to display :thinking: . I would prefer to use a boolean flag and chop `durations`. This change would be much simpler without mixing with showing the slowest tests.

lothemar commented:
> 
> 
> > Currently if you pass `--durations=0`, it'll show you **only** the timing outputs such as database creation, database cloning, and total run time, so you can only see these if you don't want to know your N-slowest tests. Passing any value greater than 0 will show you the timing outputs **and** N-slowest tests. Similarly, neither the test times aren't collected nor the `TextTest*` subclasses are used unless `durations > 0`.
> 
> Yes I know, IMO this is confusing, unexpected and out of scope of this ticket (grumbling ... 👴), maybe it's clear for others ... 🤷 , if I want to get timing outputs I have to declare the number of slowest tests to display 🤔 . I would prefer to use a boolean flag and chop `durations`. This change would be much simpler without mixing with showing the slowest tests.

We could change it so you only need to use ```--durations``` to display the timing outputs. I understand why it's confusing, and I agree it can be.

If you think changing it to ```--durations``` is still confusing, I don't mind chopping ```--durations``` and reverting back to ``--timing```.  👍

smithdc1 commented:
Here is how Pytest approaches this problem. Not saying we should copy - but someone else has given this thought (hopefully) and is the solution they landed on. This comment applies to other comments I'll make on comparisons to Pytest. 

`pytest --durations=0` — Show all times for tests and setup and teardown
`pytest --durations=1` — Just show me the slowest
`pytest --durations=10` — Slowest 10, with times, … etc.

https://medium.com/@brianokken/pytest-durations-0-show-all-times-for-tests-and-setup-and-teardown-848dccac85db

smithdc1 commented on django/test/utils.py:
```diff
@@ -841,6 +844,43 @@ def disable(self):
+            sorted_times = sorted(self.test_suite_records.items(), reverse=True, key=lambda records: records[1])
+            for key, value in sorted_times:
+                if self.n_tests > 0:
+                    results.append("%s took %0.3fs" % (key, value))
```
I'm not sure about this string. It results in the time (the most important bit) being at different indentations. I'd prefer to see it at the start so it is easier to read the timings - the test name itself is of secondary importance. 

For reference, Pytest formats strings like this. 
`0.50s call test_this.py::SlowTestCase::test_should_run_slow`

lothemar replied:
That's a good idea. I'll change it

carltongibson commented on django/test/utils.py:
```diff
@@ -174,12 +176,13 @@ def setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, paral
-                            verbosity=verbosity,
-                            keepdb=keepdb,
-                        )
+                    with time_keeper.timed("Database cloning"):
```
Can we add the alias to this string, otherwise the output is a but meh...: 

```
Database cloning took 2.384s
Database cloning took 2.398s
```

carltongibson commented on docs/releases/3.2.txt:
```diff
@@ -304,6 +304,10 @@ Tests
   :py:mod:`faulthandler` by default. This can be disabled by using the
   :option:`test --no-faulthandler` option.
 
+* Django now tracks timings like database creation, database teardown,
+  database cloning, total run time, and N slowest tests. This can be enabled by
+  using the :option:`test --timing` option. 
+
```
This isn't right. The option isn't exposed to the `test` command. 

```
(master)$ ./manage.py test --timing
usage: manage.py test [-h] [--noinput] [--failfast] [--testrunner TESTRUNNER]
                      [-t TOP_LEVEL] [-p PATTERN] [--keepdb] [-r]
                      [--debug-mode] [-d] [--parallel [N]] [--tag TAGS]
                      [--exclude-tag EXCLUDE_TAGS] [--pdb] [-b]
                      [--no-faulthandler] [-k TEST_NAME_PATTERNS] [--version]
                      [-v {0,1,2,3}] [--settings SETTINGS]
                      [--pythonpath PYTHONPATH] [--traceback] [--no-color]
                      [--force-color]
                      [test_label [test_label ...]]
manage.py test: error: unrecognized arguments: --timing
```

We're not adding `N` here currently right? 

If this is just for `runtests.py` I'm not sure we need to doc it...

carltongibson replied:
I think that given that we're here already, exposing this to `django-admin test` is worthwhile, so argument, docs, and tests for that.

carltongibson replied:
But let's save `N` for another day.

lothemar replied:
Agreed

carltongibson commented on django/test/utils.py:
```diff
@@ -841,6 +844,30 @@ def disable(self):
         setattr(Options, 'default_apps', self.old_apps)
 
 
+class time_keeper:
```
Why not `TimeKeeper`? 

We instantiate this later: `self.time_keeper = time_keeper(self.timing)`, which could equally be `self.time_keeper = TimeKeeper(self.timing)`

lothemar replied:
I thought it should be snake_case since users might want to use the class themselves to time their own test runner's operations or time something else entirely i.e it's a documented interface. Reading [PEP8](https://www.python.org/dev/peps/pep-0008/#class-names) again though does show it should be camelCase instead since it isn't going to be primarily used as a callable

carltongibson commented on django/test/utils.py:
```diff
@@ -152,7 +153,8 @@ def teardown_test_environment():
 
-def setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, parallel=0, aliases=None, **kwargs):
+def setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, parallel=0, aliases=None,
+                    time_keeper=None, **kwargs):
```
`time_keeper` can't be `None`.

carltongibson commented on django/test/utils.py:
```diff
@@ -841,6 +844,30 @@ def disable(self):
+            self.append(name, end_time)
+
+    def results(self):
+        if self.timing:
```
It seems that the `timing` parameter is doing too much work. We're storing it in the runner, plus passing it down to the time keeper class, for it only to be used here. 

Q: why is it the time keeper's job to choose whether output is displayed by the runner? (A: it's not)

I think two classes would be better than the conditional. `TimeKeeper` and `NullTimeKeeper`, then in the runner we don't store `timing` but just do: 

```
self.time_keeper = TimeKeeper() if timing else NullTimeKeeper()
```

`NullTimeKeeper` should implement no-ops for `timed()`, `append()` and `results()` and just a single `yield` for the context manager.

carltongibson commented:
One more comment here, on the slowest N proposal... — I'm minded to accept that: lots of people use `django-admin test` and showing your slowest tests is a nice option. 

* I'm a bit meh about copying `durations`. For me, `timing` is nice. If you change tool, you change tool. (But we can bikeshed for cycle.) 
*  I'd like to be able to do `--timing` and just get the current output (no slowest `N`) Could we re-use the same flag, defaulting to `0`, or leverage the verbosity settings like @felixxm mentioned... (That last probably works: pass verbosity to `TimeKeeper(verbosity)`... 

Either way, let's handle it separately: the basic feature is what we need here, and that'll help for testing the parallel runner PRs. 👍

lothemar commented:
I've added in the ```Creating``` and ```Cloning``` steps. I had to switch to a dictionary to handle the duplicate entries and maintain order of entry. I've also added the relevant docs and tests.

carltongibson commented on tests/runtests.py:
```diff
@@ -508,6 +510,11 @@ def paired_tests(paired_test, options, test_labels, parallel, start_at, start_af
         help='Discard output of passing tests.',
     )
+    parser.add_argument(
+        '-t', '--timing', action='store_true',
```
Can we drop the `-t`, I'm not convinced it's needed and the short help just puts `-t` which is opaque (if you misspell the option).

pope1ni commented on django/core/management/commands/test.py:
```diff
@@ -49,8 +49,10 @@ def add_arguments(self, parser):
     def handle(self, *test_labels, **options):
         TestRunner = get_runner(settings, options['testrunner'])
 
+        run_keeper = TimeKeeper() if options["timing"] else NullTimeKeeper()
```
`run_keeper` → `time_keeper` for consistency? It feels like a hangover from when the class was called `time_keeper` before it changed to `TimeKeeper`.

Are we also not still using single quotes?

```suggestion
        time_keeper = TimeKeeper() if options['timing'] else NullTimeKeeper()
```

pope1ni commented on django/core/management/commands/test.py:
```diff
@@ -49,8 +49,10 @@ def add_arguments(self, parser):
         test_runner = TestRunner(**options)
-        failures = test_runner.run_tests(test_labels)
-
+        with run_keeper.timed("Total run"):
+            failures = test_runner.run_tests(test_labels)
+        run_keeper.results()
```
```suggestion
        with time_keeper.timed('Total run'):
            failures = test_runner.run_tests(test_labels)
        time_keeper.results()
```

pope1ni commented on django/test/runner.py:
```diff
@@ -704,7 +712,8 @@ def run_tests(self, test_labels, extra_tests=None, **kwargs):
         suite = self.build_suite(test_labels, extra_tests)
         databases = self.get_databases(suite)
-        old_config = self.setup_databases(aliases=databases)
+        with self.time_keeper.timed("Total database setup"):
```
```suggestion
        with self.time_keeper.timed('Total database setup'):
```

pope1ni commented on django/test/runner.py:
```diff
@@ -714,13 +723,15 @@ def run_tests(self, test_labels, extra_tests=None, **kwargs):
         finally:
             try:
-                self.teardown_databases(old_config)
+                with self.time_keeper.timed("Total database teardown"):
```
```suggestion
                with self.time_keeper.timed('Total database teardown'):
```

pope1ni commented on django/test/utils.py:
```diff
@@ -841,6 +845,37 @@ def disable(self):
 
+class TimeKeeper:
+    def __init__(self):
+        self.records = {}
```
```suggestion
        self.records = collections.defaultdict(list)
```

carltongibson replied:
This alters the behavour. See https://github.com/django/django/pull/13224#pullrequestreview-462400381 🤔

pope1ni replied:
Ah. It is because we are pre-populating `self.records` before the `yield` in the context manager. Changing to `defaultdict` means that the key only gets added after when the context manager is exiting and we do `self.records[name].append(end_time)`.

This means that nested uses of `time_keeper.timed()` - as we do in the database setup -- end up being out of order.

pope1ni commented on django/test/utils.py:
```diff
@@ -841,6 +845,37 @@ def disable(self):
+    @contextmanager
+    def timed(self, name):
+        start_time = time.perf_counter()
+        if name not in self.records.keys():
+            self.records[name] = []
```
```suggestion
```

pope1ni commented on django/test/utils.py:
```diff
@@ -841,6 +845,37 @@ def disable(self):
+            end_time = time.perf_counter() - start_time
+            self.records[name].append(end_time)
+
+    def results(self):
```
Should we name this `print_results` or move the `sys.stderr.write()` call outside?
(I don't know if there would be a use for returning the generated text for something other than printing it out.)

carltongibson commented:
@lothemar Looks like https://github.com/django/django/pull/13224/commits/c474b4a82205a1467d6bf7dddb1a3cfd0ec41720 is necessary. I was thinking not. Didn't think through yet if that's the cleanest, so can you think about it. Ta!

pope1ni commented on docs/topics/testing/advanced.txt:
```diff
@@ -577,6 +580,7 @@ execute and tear down the test suite.
     .. versionadded:: 3.2
 
         The ``enable_faulthandler`` argument was added.
+        The ``timing`` argument was added.
```
```suggestion
        The ``enable_faulthandler`` and ``timing`` arguments were added.
```

pope1ni commented on tests/runtests.py:
```diff
@@ -508,6 +510,11 @@ def paired_tests(paired_test, options, test_labels, parallel, start_at, start_af
     )
+    parser.add_argument(
+        '--timing', action='store_true',
+        help='Output timings such as DB creation, DB cloning, DB teardown and'
+        ' total test run time.',
```
Prefer "database" to "DB"?

```suggestion
        help='Output timings such as database creation, cloning, and teardown;'
        ' and total test run time.',
```

pope1ni commented on tests/runtests.py:
```diff
@@ -568,13 +575,17 @@ def paired_tests(paired_test, options, test_labels, parallel, start_at, start_af
-            getattr(options, 'test_name_patterns', None),
-            options.start_at, options.start_after, options.pdb, options.buffer,
-        )
+        run_keeper = TimeKeeper() if options.timing else NullTimeKeeper()
+        with run_keeper.timed("Total run"):
```
```suggestion
        time_keeper = TimeKeeper() if options.timing else NullTimeKeeper()
        with time_keeper.timed('Total run'):
```

pope1ni commented on tests/runtests.py:
```diff
@@ -568,13 +575,17 @@ def paired_tests(paired_test, options, test_labels, parallel, start_at, start_af
+                options.start_at, options.start_after, options.pdb, options.buffer,
+                options.timing
+            )
+        run_keeper.results()
```
```suggestion
        time_keeper.results()
```

pope1ni commented on django/test/utils.py:
```diff
@@ -841,6 +845,37 @@ def disable(self):
+        self.records = {}
+
+    @contextmanager
+    def timed(self, name):
```
Also not sure if there is a better name than `timed` here? Perhaps `measure`?

carltongibson commented:
Also, on the extra test case: I guess we'd need to mock out the test runner -- some horrible regress otherwise 😳

carltongibson commented on django/test/runner.py:
```diff
@@ -525,6 +528,11 @@ def add_arguments(cls, parser):
             '--no-faulthandler', action='store_false', dest='enable_faulthandler',
             help='Disables the Python faulthandler module during tests.',
         )
+        parser.add_argument(
+            '--timing', action='store_true',
+            help='Output timings such as DB creation, DB cloning, DB teardown and'
+            ' total test run time.',
+        )
```
This should match the updated version in `runtests.py`.

lothemar commented:
Sorry for the relative silence on this ticket. I had an issue with my ISP (it has been resolved now). I'm pushing a patch tonight on this ticket.

Concerning adding a test for the option's output, I think it's reasonable and should be there to detect regressions. I'll add it and push it.

The ```defaultdict``` optimization is beautiful! It'll be easy to integrate into the current patch.

felixxm commented on django/test/runner.py:
```diff
@@ -525,6 +528,11 @@ def add_arguments(cls, parser):
         )
+        parser.add_argument(
+            '--timing', action='store_true',
+            help='Output timings such as database creation, cloning, and teardown;'
+            ' and total test run time.',
```
I'm not sure why we have here a semicolon :thinking:, please use also hanging indentation:
 ```suggestion
            help=(
                'Output timings such as database creation, cloning, teardown, '
                ' total test run time.',
            ),
```

pope1ni replied:
That came from my suggestion. Originally it was `DB creation, DB cloning, DB teardown and total test run time` and, aside from `DB` → `database`, it seemed unnecessary to repeat `database` multiple times.

Semicolons can be used in complex lists where there are commas used within the list items themselves. I perceived there to be two list items here:

- database creation, cloning, and teardown
- total test run time

Your suggested wording now no longer makes it clear that "cloning" and "teardown" are database-related.

felixxm commented on django/test/utils.py:
```diff
@@ -152,7 +154,8 @@ def teardown_test_environment():
 
 
-def setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, parallel=0, aliases=None, **kwargs):
+def setup_databases(verbosity, interactive, time_keeper, keepdb=False, debug_sql=False, parallel=0,
+                    aliases=None, **kwargs):
```
Use hanging indentation.

felixxm replied:
Adding `time_keeper` as the 3rd argument can cause issues for people that rely on the current signature and do not use keyword args.

felixxm commented on docs/releases/3.2.txt:
```diff
@@ -319,6 +319,10 @@ Tests
   :py:mod:`faulthandler` by default. This can be disabled by using the
   :option:`test --no-faulthandler` option.
 
+* Django now tracks timings like database creation, database teardown,
```
```
* The new :option:`test --timing` option allows tracking ...
```

felixxm commented on tests/runtests.py:
```diff
@@ -508,6 +510,11 @@ def paired_tests(paired_test, options, test_labels, parallel, start_at, start_af
     )
+    parser.add_argument(
+        '--timing', action='store_true',
+        help='Output timings such as database creation, cloning, and teardown;'
+        ' and total test run time.',
```
Ditto

felixxm commented on tests/runtests.py:
```diff
@@ -568,13 +575,17 @@ def paired_tests(paired_test, options, test_labels, parallel, start_at, start_af
+                options.exclude_tags,
+                getattr(options, 'test_name_patterns', None),
+                options.start_at, options.start_after, options.pdb, options.buffer,
+                options.timing
```
Add trailing comma.

carltongibson commented on docs/topics/testing/advanced.txt:
```diff
@@ -576,7 +579,7 @@ execute and tear down the test suite.
     .. versionadded:: 3.2
 
-        The ``enable_faulthandler`` argument was added.
+        The ``enable_faulthandler`` and ``timing`` arguments were added.
```
I'm on that.

felixxm commented:
> How would you want to handle that?

Maybe we can change keywords arguments to keyword-only arguments and document this :thinking: Also we need to update `django.test.utils.setup_databases()` docs in `topics/testing/advanced.txt`.

lothemar commented on django/test/utils.py:
```diff
@@ -152,7 +154,8 @@ def teardown_test_environment():
 
 
-def setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, parallel=0, aliases=None, **kwargs):
+def setup_databases(verbosity, interactive, *, time_keeper, keepdb=False, debug_sql=False, parallel=0,
```
@felixxm Changed it to keyword-only and added the relevant docs. Let me know if there are any missing changes or incorrect changes

felixxm replied:
It looks that you missed some of Carlton's edits, e.g. https://github.com/django/django/pull/13224#discussion_r468468374.

felixxm commented on docs/topics/testing/advanced.txt:
```diff
@@ -659,11 +662,15 @@ Methods
     Returns a ``TestSuite`` instance ready to be run.
 
-.. method:: DiscoverRunner.setup_databases(**kwargs)
+.. method:: DiscoverRunner.setup_databases(verbosity, interactive, *, **kwargs)
```
`DiscoverRunner.setup_databases()` signature has not been changed, please revert this and a `versionadded` annotation.

carltongibson replied:
Yep, good. 👍

adamchainz commented:
👍👍🐙

laymonage commented on docs/ref/django-admin.txt:
```diff
@@ -1523,6 +1523,12 @@ Django automatically calls :func:`faulthandler.enable()` when starting the
+
+.. versionadded:: 3.2
+
+Outputs timings, including including database setup and total run time.
```
Minor typo here :)

felixxm replied:
Thanks, #13318.
